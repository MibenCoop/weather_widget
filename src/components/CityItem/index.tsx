import React from "react";

import { City } from "../../interfaces";

export interface CityProps {
  city: City;
}

const CityItem: any = ({ city }: CityProps) => {
  // console.log("city", city);
  return <li style={{ textAlign: "left" }}>{city.city}</li>;
};

export default CityItem;
