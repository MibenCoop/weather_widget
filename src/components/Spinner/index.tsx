import React from "react";

const Spinner = () => <div>Waiting...</div>;

export default Spinner;
