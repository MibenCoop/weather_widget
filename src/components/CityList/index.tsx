import React from 'react';

import CityItem from '../CityItem';
import { City } from '../../interfaces';

export interface Props {
    cities: Array<any>;
}

const CityList: any = ({ cities }: Props) =>
    cities.map((item: City) => {
        // console.log("item", item);
        return <CityItem key={item.lat + item.lng} city={item} />;
    });

export default CityList;
