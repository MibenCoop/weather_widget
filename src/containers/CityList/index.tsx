import React, { Component, useState, useEffect } from 'react';
import { connect } from 'react-redux';

import { fetchCities } from '../../actions/cities';
import { increment } from '../../actions/counter';
import CitiesSelector from '../../selectors/citySelector';

import Spinner from '../../components/Spinner';
import CityList from '../../components/CityList';
import { SignupForm } from '../../components/SignupForm';
import Counter from '../../components/Counter';

import { City, CityListState, SelectedCities } from '../../interfaces';

interface Props {
    cities: Array<object>;
    biggestCities: Array<object>;
    counter: number;
    isFetching: boolean;
    fetchCities: Function;
}

interface State {
    playOrPause?: string;
    selectedCities?: SelectedCities;
    hasSelectedCities: boolean;
}

export function CityListContainer(props: Props) {
    const { fetchCities, biggestCities, cities, isFetching } = props;
    const [hasSelectedCities, setHasSelectedCities] = useState(false);
    const [selectedCities, setSelectedCities] = useState([]);
    useEffect(() => {
        const fetchCityList = () => {
            fetchCities();
        };
        return fetchCityList();
    }, []);

    const selectCities = ({ city, minTemperature, maxTemperature }: SelectedCities) => {
        const selectedCitiesBySelector = getCitiesByCelector(city, minTemperature, maxTemperature);
        setHasSelectedCities(true);
        setSelectedCities(selectedCitiesBySelector);
    };

    const getCitiesByCelector = (name: any, minTemperature: any, maxTemperature: any) => {
        const { cities }: any = props;
        if (name.length > 0) {
            return cities.filter((city: any) => city.name === name);
        } else if (minTemperature && maxTemperature) {
            return cities.filter(
                (city: any) => city.temperature <= maxTemperature && city.temperature >= minTemperature,
            );
        } else if (minTemperature) {
            return cities.filter((city: any) => city.temperature >= minTemperature);
        } else if (maxTemperature) {
            return cities.filter((city: any) => city.temperature <= maxTemperature);
        }
    };

    if (isFetching || !cities) return <Spinner />;
    return (
        <>
            <SignupForm selectCities={(values: SelectedCities) => selectCities(values)} />
            {hasSelectedCities ? (
                <>
                    <p>Selected cities</p>
                    <CityList cities={selectedCities} />
                </>
            ) : null}
            <p>Top ten biggest City in Russia</p>
            <CityList cities={biggestCities} />
            <p>All</p>
            <CityList cities={cities} />
        </>
    );
}

const mapStateToProps = ({ cities, counter }: CityListState) => {
    return {
        cities: cities.cities,
        isFetching: cities.isFetching,
        biggestCities: CitiesSelector(cities.cities),
        counter: counter,
    };
};

export default connect(mapStateToProps, {
    fetchCities: fetchCities,
    increment: increment,
})(CityListContainer);
