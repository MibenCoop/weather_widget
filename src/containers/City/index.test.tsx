import React from 'react';

import { shallow, mount } from 'enzyme';

import { CityWeatherList } from './index';
import Spinner from '../../components/Spinner';
import CityList from '../../components/CityList';
import CityItem from '../../components/CityItem';
import { Provider } from 'react-redux';

const props = {
    name: 'Moscow',
    temparature: 10,
    isFetching: false,
    fetchWeather: jest.fn(),
};

const cityItemProps = {
    name: 'Moscow',
    temparature: 10,
};

describe('City Weather List container', () => {
    const cityList = shallow(<CityWeatherList {...props} />);

    it('should match snapshot', () => {
        expect(cityList).toMatchSnapshot();
    });
    it('should render prop name correctly', () => {
        expect(cityList.prop('name')).toEqual(props.name);
    });
    it('should fetch weather on page loading', () => {
        expect(props.fetchWeather).toBeCalledTimes(1);
    });
    it('should be called data fetching if component will be updated', () => {
        //force a rerender
        cityList.setProps({ name: 'Rostov' });
        expect(cityList.prop('name')).toEqual('Rostov');
        expect(props.fetchWeather).toBeCalledTimes(2);
    });
    it("should not render spinner if doesn't fetching", () => {
        cityList.setProps({ isFetching: false });
        expect(cityList.find(Spinner).length).toEqual(0);
    });
    it('should not render city item if fetching', () => {
        cityList.setProps({ isFetching: true });
        expect(cityList.find(CityItem).length).toEqual(0);
    });
    it('should render spinner if fetching data', () => {
        cityList.setProps({ isFetching: true });
        expect(cityList.find(Spinner).length).toEqual(1);
    });
});
