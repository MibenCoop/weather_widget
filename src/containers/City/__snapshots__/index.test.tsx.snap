// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City Weather List container should match snapshot 1`] = `
<City
  name="Moscow"
  temperature={10}
/>
`;
